src/arithmetic/rewrite_simplify.cc

```diff
@@ -0,0 +1,646 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file rewrite_simplify.cc
+ * \brief Rewrite-rule based simplification.
+ */
+// Acknowledgement: Most rewrite-rules are from Halide.
+#include <tvm/arithmetic.h>
+#include <tvm/expr_operator.h>
+#include <tvm/ir_mutator.h>
+#include "const_fold.h"
+#include "pattern_match.h"
+
+namespace tvm {
+namespace arith {
+
+using namespace ir;
+
```

Comment:
In my opinion, this approach to specifying rewriting rules has pros and cons. The good thing is that rule definition looks really clear. The bad thing I see is that rules are declared one after another in a single C++ function. We can't combine rules into bigger rules, and, more importantly, we can't call them individually, for example in order to apply a testcase.

Reply:
in theory, CSE(common subexpression elimination) in compiler could pull out some of the common computed element of these rules and not recompute them